rmw_connext_shared_cpp/src/node_names.cpp

```diff
@@ -90,23 +97,40 @@ get_node_names(
         }
       }
     }
+
+    // ignore discovered participants without a name
     if (name.empty()) {
-      // ignore discovered participants without a name
-      node_names->data[i] = nullptr;
       continue;
     }
-    node_names->data[i] = rcutils_strdup(name.c_str(), allocator);
-    if (!node_names->data[i]) {
+
+    tmp_names_list.data[named_nodes_num] = rcutils_strdup(name.c_str(), allocator);
+
+    if (!tmp_names_list.data[named_nodes_num]) {
       RMW_SET_ERROR_MSG("could not allocate memory for node name")
-      rcutils_ret = rcutils_string_array_fini(node_names);
+      rcutils_ret = rcutils_string_array_fini(&tmp_names_list);
       if (rcutils_ret != RCUTILS_RET_OK) {
         RCUTILS_LOG_ERROR_NAMED(
           "rmw_connext_cpp",
           "failed to cleanup during error handling: %s", rcutils_get_error_string_safe())
       }
       return RMW_RET_BAD_ALLOC;
     }
+
+    ++named_nodes_num;
+  }
+
+  rcutils_ret = rcutils_string_array_init(node_names, named_nodes_num, &allocator);
```

Comment:
since we initialise this here with the required size, we can remove [this previous initialisation](https://github.com/ros2/rmw_connext/pull/303/files#diff-bccfea1954134a50cf75b01ae95188e3R56). Removing it above also means we won't have to fini the `node_names` array in the various error cases.